pkg/gardenlet/controller/shoot/shoot/cleaner.go

```diff
@@ -0,0 +1,204 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shoot
+
+import (
+	"context"
+	"time"
+
+	machinev1alpha1 "github.com/gardener/machine-controller-manager/pkg/apis/machine/v1alpha1"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	resourcesv1alpha1 "github.com/gardener/gardener/pkg/apis/resources/v1alpha1"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/extensions"
+	"github.com/gardener/gardener/pkg/utils/flow"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+	utilclient "github.com/gardener/gardener/pkg/utils/kubernetes/client"
+)
+
+const (
+	// defaultInterval is the default interval for retry operations.
+	defaultInterval = 5 * time.Second
+	// defaultTimeout is the default timeout for waiting for resources to be migrated or deleted.
+	defaultTimeout = 5 * time.Minute
+)
+
+var (
+	extensionKindToObjectList = map[string]client.ObjectList{
+		extensionsv1alpha1.BastionResource:               &extensionsv1alpha1.BastionList{},
+		extensionsv1alpha1.ContainerRuntimeResource:      &extensionsv1alpha1.ContainerRuntimeList{},
+		extensionsv1alpha1.ControlPlaneResource:          &extensionsv1alpha1.ControlPlaneList{},
+		extensionsv1alpha1.DNSRecordResource:             &extensionsv1alpha1.DNSRecordList{},
+		extensionsv1alpha1.ExtensionResource:             &extensionsv1alpha1.ExtensionList{},
+		extensionsv1alpha1.InfrastructureResource:        &extensionsv1alpha1.InfrastructureList{},
+		extensionsv1alpha1.NetworkResource:               &extensionsv1alpha1.NetworkList{},
+		extensionsv1alpha1.OperatingSystemConfigResource: &extensionsv1alpha1.OperatingSystemConfigList{},
+		extensionsv1alpha1.WorkerResource:                &extensionsv1alpha1.WorkerList{},
+	}
+
+	mcmKindToObjectList = map[string]client.ObjectList{
+		"MachineDeployment": &machinev1alpha1.MachineDeploymentList{},
+		"MachineSet":        &machinev1alpha1.MachineSetList{},
+		"MachineClass":      &machinev1alpha1.MachineClassList{},
+		"Machine":           &machinev1alpha1.MachineList{},
+	}
+)
+
+type cleaner struct {
+	seedClient       client.Client
+	gardenClient     client.Client
+	seedNamespace    string
+	projectNamespace string
+	backupEntryName  string
+	log              logr.Logger
+}
+
+// NewCleaner creates a cleaner with the given clients and logger, for a shoot with the given namespaces and backupEntryName.
+func NewCleaner(seedClient, gardenClient client.Client, seedNamespace, projectNamespace string, backupEntryName string, log logr.Logger) *cleaner {
+	return &cleaner{
+		seedClient:       seedClient,
+		gardenClient:     gardenClient,
+		seedNamespace:    seedNamespace,
+		projectNamespace: projectNamespace,
```

Comment:
Where is this `projectNamespace` used?

Reply:
This is leftover from the time when we manually deleted backupEntry, removed.